### PR TITLE
nixos/certmgr: init

### PR DESCRIPTION
**Pull request by johanot**
###### Motivation for this change

Added nixos module for easy configuration of certmgr daemon service.
See #44406 for reference.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by infinisil — CHANGES_REQUESTED** [review]
Here I am again :)

**Comment by Infinisil** [comment]
Also: Can this in any way interact with ACME? NixOS has a great ACME module, maybe something can be done with that.

**Comment by johanot** [comment]
@Infinisil Let me start with the general question about acme. I like acme and Let's Encrypt very much myself, and I use it everywhere I can. I can think of only one (maybe two) scenarios where using acme is not possible.

1) When you don't have access to the internet at all, or you just don't have any means of responding to acme challenges.

2) When you need x509 certificates with non-DNS names in the subject section.

It's no secret that I plan to use cfssl and certmgr for a refactoring of the Kubernetes module. The Kubernetes PKI-setup in particular, requires certificates with rolenames like `system:controller-manager` and `system:scheduler` stamped in the certificate subject section, common name (CN) field. As far as I know, this is not something you can request with Let's Encrypt? 

.. and even if you could change the default system role names to something like `controller-manager.kubernetes.domain.org`, you would not be able to use Kubernetes without opening your cluster to the internet. IMHO, cluster-components should be able to communicate via TLS without depending on external services.

All that said; I really wish products like cfssl or Vault PKI had support for the acme protocol. That though, would also require a change of the NixOS acme module, AFAIK? It's hardcoded for Let's Encrypt as issuer :).

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Fixed in: https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Removed parens as part of: https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b .. Will look into the possibility (and potential problems) of merging impSpecs and declSpecs now.

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Fixed in https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Fixed in https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b

File: nixos/tests/certmgr.nix
Comment (johanot) [reply nixos/tests/certmgr.nix]: Fixed in https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b

File: nixos/tests/certmgr.nix
Comment (johanot) [reply nixos/tests/certmgr.nix]: Done. https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b

File: nixos/tests/certmgr.nix
Comment (johanot) [reply nixos/tests/certmgr.nix]: Done. https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Whoops.. c/p error. as for the description. It is changed as part of https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b.

Regarding other init-systems, yes. We probably won't need them in NixOS. But the `command` service manager is useful, when you just want to execute a verbatim command or script instead of invoking systemd, and `dummy` can be useful for debugging.

I'm personally ok with leaving the list as-is (vendor provided). But if you think so, I'm also happy to remove `circus` `openrc` and `sysv`. As long as default remains to be systemd.

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Done. https://github.com/NixOS/nixpkgs/pull/44556/commits/a94946e1505852aa7d8e0e7700af74e38c38819b

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Combined `impSpecs` and `declSpecs` in https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Removed option `mkTargetDirs` as part of https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Not entirely sure what you are saying here? Changed the assertion message slightly here: https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2 .. The point is that setting `auth_key` would certainly place a secret in nix-store, which we want to prevent. `auth_key_file` can be safely used, if set to a non-store-path.

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Changed as part of https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2 .. Still need to distinguish specs of type attrset and specs of type path, since the former needs to be written to the store before linking.

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Changed as part of https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2. Using `filterAttrsRecursive`. Escaping shell-args. But still using nix for `dirOf`and `unique`, since I don't see the reason for doing this at runtime (in bash)?

Tell me what you think about the overall solution, please :) - after merge of impSpecs and declSpecs, that is.

File: nixos/tests/certmgr.nix
Comment (johanot) [reply nixos/tests/certmgr.nix]: Done. https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2

File: nixos/modules/services/security/certmgr.nix
Comment (Infinisil) [inline nixos/modules/services/security/certmgr.nix]: This seems to be the same as just `allSpecs = pkgs.linkFarm "certmgr.d" specs;`

File: nixos/modules/services/security/certmgr.nix
Comment (Infinisil) [inline nixos/modules/services/security/certmgr.nix]: Could also use this which looks a bit cleaner imo:
```nix
specs = mapAttrsToList (n: v: rec {
  name = n + ".json";
  path = if isAttrs v then pkgs.writeText name (builtins.toJSON v) else v;
}) cfg.specs;
```

File: nixos/modules/services/security/certmgr.nix
Comment (Infinisil) [inline nixos/modules/services/security/certmgr.nix]: Couple things: `n` isn't used -> `concatMap/map` instead. More descriptive name instead of `v` (can be confusing with later `v`). Named this variable properly, and removed the flexibility of `set` and using `cfg.specs` always instead, that's all we need. I don't think using `unique` is worth it, it's an expensive operation (O(n^2)) and mkdir handles duplicates just fine. And some more things. All in all I think it should look like this:

```nix
specPaths = map dirOf (concatMap (spec:
  if isAttrs spec then
    collect isString (filterAttrsRecursive (n: v: isAttrs v || n == "path") spec)
  else
    [ spec ]
) (attrValues cfg.specs)));

preStart = ''
  ${concatStringsSep " \\\n" (["mkdir -p"] ++ map escapeShellArg specPaths)}
  ...
''
```

File: nixos/modules/services/security/certmgr.nix
Comment (Infinisil) [inline nixos/modules/services/security/certmgr.nix]:
```nix
mapAttrsToList (_: value: value) cfg.specs = attrValues cfg.specs
```

File: nixos/tests/certmgr.nix
Comment (Infinisil) [inline nixos/tests/certmgr.nix]: Why this weird check? Maybe a `$machine->sleep(3)` will work instead if the `waitForUnit` really isn't enough.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/tests/certmgr.nix
Comment (johanot) [reply nixos/tests/certmgr.nix]: The check is there to confirm that nginx is in fact restarted by certmgr after certificate issue. `Starting nginx ...` would be present in the journal more than once if nginx was restarted. `waitForUnit` will terminate just fine even if certmgr restart action is set to no-op/dummy.

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Fixed in https://github.com/NixOS/nixpkgs/pull/44556/commits/9082a6152a3bc2c27942908b529b05ab10efd224 .. And I added an extra assertion to check if the list of specs is empty.

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Changed to match your suggestion in: https://github.com/NixOS/nixpkgs/pull/44556/commits/9082a6152a3bc2c27942908b529b05ab10efd224

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Changed to match your suggestion in: https://github.com/NixOS/nixpkgs/pull/44556/commits/9082a6152a3bc2c27942908b529b05ab10efd224

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Of course. I just forgot to reduce it after all my testing. Fixed in https://github.com/NixOS/nixpkgs/pull/44556/commits/9082a6152a3bc2c27942908b529b05ab10efd224

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: `{ ... }` removed in https://github.com/NixOS/nixpkgs/pull/44556/commits/9082a6152a3bc2c27942908b529b05ab10efd224

Comment (srhb) [reply nixos/modules/services/security/certmgr.nix]: Should the check for `path` really be `(builtins.tryEval (builtins.substring 0 1 (toString x) == "/")).value`? It feels icky to use `tryEval` here, but I certainly do expect `either` to not care about the order of its arguments, and `path` does break this as described.

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: @Infinisil Finally I can answer this question. Bash is needed for certmgr when svcManager in in `command`-mode. See: https://github.com/cloudflare/certmgr#command-svcmgr-and-how-to-use-it

I've created a new test case for command-mode here: https://github.com/NixOS/nixpkgs/pull/44556/commits/27e6197ed9a4907a0cb5fa80f3ba03868314dd14

Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: .. and the answer is no. It doesn't work without bash explicitly included in PATH.

File: nixos/modules/services/security/certmgr.nix
Comment (srhb) [reply nixos/modules/services/security/certmgr.nix]: While I understand the intention behind combining the specs, I think it'll lead to a poorer understanding of the intended separation. It's important that `impSpec`s are never imported into the store, since they'll most likely (always?) contain secrets. The `either` machinery necessary to facilitate a joined `specs` introduces a lot of opportunities for future maintainers to accidentally import the paths -- and even if it's safe (which it is in the current state) the error messages if you muck up the value are not great.

In contrast, `impSpecs` might look non-dry, but it requires very trivial machinery and it'll be easier to ensure that none of its values are accidentally leaked into the store.

File: nixos/modules/services/security/certmgr.nix
Comment (srhb) [reply nixos/modules/services/security/certmgr.nix]: Disregard my comment here, it was pointed out to me that it's complete nonsense, since string coercion errors aren't catchable with tryEval.

File: nixos/tests/certmgr.nix
Comment (Infinisil) [reply nixos/tests/certmgr.nix]: Ah I see

**Comment by Infinisil** [comment]
Regarding the thing mentioned by @srhb in https://github.com/NixOS/nixpkgs/pull/44556#discussion_r208901272 we had a small discussion about it: https://logs.nix.samueldr.com/nixos-dev/2018-08-09#1533829782-1533831340;

I personally like to have this additional complexity of having path and option specs combined. I feel it makes more sense, since a spec of a specific name can be one of either, which corresponds to this `name -> path/attrset` mapping.

But it also has the potential for messing up the implementation (the `isAttrs` distinguishing), which I think is also important to get right, especially with security related things, nothing sensitive should land in the store and such.

An actual *good* solution would be to have ADT's in the module system (see @shlevy's https://github.com/shlevy/nix-adt, which I think doesn't fully fit into the module system however), but that's probably a bit off as of now.

@johanot What do you think of this? I'm good with either, keeping the options for path/attrsets together or splitting them.

When that's cleared up, this PR looks about ready for merging to me :)

File: nixos/modules/services/security/certmgr.nix
Comment (Infinisil) [inline nixos/modules/services/security/certmgr.nix]: Almost missed this, but very important:

Like this, this service is running as root, which should be avoided (also see #41092). Check out the `DynamicUser` and `RuntimeDirectory` systemd options, they should probably suffice.

File: nixos/modules/services/security/certmgr.nix
Comment (Infinisil) [inline nixos/modules/services/security/certmgr.nix]: `lib` is already `with;`'d above, can be removed.

File: nixos/modules/services/security/certmgr.nix
Comment (Infinisil) [inline nixos/modules/services/security/certmgr.nix]:
```nix
any (e: e == x) [ ... ] = elem x [ ... ]
```

the parens around that can also be removed, function application binds very strongly

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/security/certmgr.nix
Comment (johanot) [reply nixos/modules/services/security/certmgr.nix]: Unfortunately it is not as simple as `DynamicUser`.. Ok.. I looked at the acme module, curious to see how it handles private key file-permissions, when running as non-root. Certmgr has to be able to issue certificates, and generate private keys for daemon A, B and C - each running with different users - and still private key files must be protected and readable for the exact daemon it is intended to only.

It looks like the acme module creates a systemd service for each cert, allowing for different users per cert. @Infinisil .. Do you think I should rewrite the certmgr module along the same lines, or do you have other suggestions?

**Comment by Infinisil** [comment]
Oh I see, well if that's the case then it's probably fine to use root, acme uses root by default as well. I'm fine with leaving it like that. No need for additional complexity, unless somebody really needs it.

**Comment by johanot** [comment]
@Infinisil I removed the `with lib` and replaced `any (e == x)` with `elem x`. And then squashed. Feel free to merge if you have no more comments.

**Comment by srhb** [comment]
@grahamcofborg test certmgr.systemd certmgr.command

**Review by infinisil — APPROVED** [review]

**Comment by Infinisil** [comment]
Ran both tests locally, both successfully. Thanks for the PR! I'll merge in a bit.

**Pull request merged**
